src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -306,28 +374,57 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
   bool supports_streams() const noexcept override { return resource->supports_streams(); }
 
   /**
-   * Update the internal state so that a specific thread is associated with a task.
+   * Update the internal state so that a specific thread is dedicated to a task.
    * This may be called multiple times for a given thread and if the thread is already
-   * associated with the task, then most of the time this is a noop. The only exception
+   * dedicated to the task, then most of the time this is a noop. The only exception
    * is if the thread is marked that it is shutting down, but has not completed yet.
    * This should never happen in practice with Spark because the only time we would
    * shut down a task thread on a thread that is different from itself is if there
    * was an error and the entire executor is shutting down. So there should be no
    * reuse.
    */
-  void associate_thread_with_task(long thread_id, long task_id)
+  void start_dedicated_task_thread(long thread_id, long task_id)
   {
     std::unique_lock<std::mutex> lock(state_mutex);
     if (shutting_down) { throw std::runtime_error("spark_resource_adaptor is shutting down"); }
+    auto found = threads.find(thread_id);
+    if (found != threads.end()) {
+      if (found->second.task_id >= 0 && found->second.task_id != task_id) {
+        if (is_log_enabled) {
+          std::stringstream ss;
+          ss << "desired task_id " << task_id;
+
+          log_status("FIXUP", thread_id, found->second.task_id,
+                  found->second.state, ss.str().c_str());
```

Comment:
The pattern `ss.str().c_str()` is used a lot here and can cause undefined behavior/crash when logging is enabled. That is because `.str().c_str()` is a pointer to a temporary string produced by `.str()` which can be destroyed immediately.
```suggestion
         auto const log_str = ss.str();
          log_status("FIXUP", thread_id, found->second.task_id,
                  found->second.state, log_str.c_str());
```

Reply:
Another option is, for short string message like this, just use `std::string` directly:
```
auto const log_str = std::string("desired task_id ") + std::string(task_id);
          log_status("FIXUP", thread_id, found->second.task_id,
                  found->second.state, log_str.c_str());

```